### DISCUSSION TITLE
Docs explaining significance/mechanics of parser state

### DISCUSSION THREAD

**Discussion by smackesey** [tree-sitter/tree-sitter#769] [Q&A] [answered]
This post is really more of a question, but since I don't see it explained in the docs, I also think it merits an issue request for documentation.

I'm writing a parser that uses an external scanner. I just debugged a situation in which I seemed to be experiencing non-deterministic parsing. That is, parsing the same string would sometimes work and sometimes produce an error. I was able to eliminate this problem by always using a fresh parser object (if it matters, I'm using `node-tree-sitter`). This tells me that the operative factor was something in the state of the `Parser`.

However, I'm having a hard time trying to understand the problem. The docs don't provide much information.

The `Using Parsers` section just says: "A TSParser is a stateful object that can be assigned a TSLanguage and used to produce a TSTree based on some source code." The nature of this state is not elaborated on.

The `Creating Parsers` section is more helpful. It describes the functions you must write for your external scanner, but it says that the buffer "will ultimately be stored in the syntax tree". I read this as the parser state defined in serialize/deserialize only being important for applying edits to the same tree. But my experience above suggests this is not the case-- that is, the parser state somehow affects a fresh parse of a string with no existing tree.

So a  few questions that I think the docs could benefit from answering (and that I would appreciate some help on here):

- When exactly are `serialize`/`deserialize` called? The docs say that `serialize` is called after each token is resolved. Is `deserialize` called *before* attempting to resolve each token? And, if the serialized data is stored in a syntax tree, then what happens when you call `parse` on two different strings (there being no defined tree for the second string).
- Is there best practice for when to create a new Parser object and when to reuse an existing one? Are there significant performance implications to creating lots of Parser objects? My application is a language server and, as a result of my debug session, I am now creating a different parser object for each managed document. But I don't know if that is really necessary.

**Comment by maxbrunsfeld** [comment]
> Is there best practice for when to create a new Parser object and when to reuse an existing one? 

You shouldn't ever need to create a new `Parser` object, unless you need to perform more than one parse at the same time, using multiple threads. The parser does carry some state that you explicitly set on it, via methods like like `setLanguage`, `setTimeoutMicros`, `setLogger`, but that is it. Also notably, the parser does store an instance of your external scanner.

> When exactly are serialize/deserialize called? The docs say that serialize is called after each token is resolved. Is `deserialize` called before attempting to resolve each token?

Yeah, it is. And, if the serialized data is stored in a syntax tree, then what happens when you call parse on two different strings (there being no defined tree for the second string). In all cases, the `deserialize` method is called with the state that was serialized after the previous external token in the same parse.

It sounds like your `serialize`  and `deserialize` methods may not be correctly encoding and recreating your scanner's state?

**Comment by ahelwer** [comment, marked as answer]
Thought I'd add my experience here because I just experienced some nondeterministic behavior of my own - possibly different from yours. The issue was ultimately in my `deserialize` function, as Max suspected up above:
```
struct Scanner {
   ...
    void deserialize(const char* buffer, const unsigned length) {
        if (length > 0) {
            // rehydrate Scanner state variables here
            ...
        }
    }
    ...
}

extern "C" {
    ...
    void tree_sitter_languagename_external_scanner_deserialize(void* payload, const char* buffer, unsigned length) {
        Scanner* scanner = static_cast<Scanner*>(payload);
        scanner->deserialize(buffer, length);
    }
    ...
}
```
For whatever reason, occasionally tree-sitter called the `tree_sitter_languagename_external_scanner_deserialize` function with a buffer of length 0. I think tree-sitter maybe does this during error recovery, or other scenarios where it replays parsing some section of code? Unsure. Anyway, you might notice my `Scanner::deserialize` function doesn't do anything if it is passed a buffer of length 0. You might *also* notice that the `Scanner*` pointer seems to point to a previously-allocated Scanner object. You should be able to figure it out from here - when tree-sitter calls `deserialize` with a previously-allocated `Scanner` object and buffer of length 0, instead of resetting the state of the `Scanner` object the `Scanner` state just remains the same as *whatever state it was previously* - which could be all sorts of weird things. The fix was very simple:
```
void deserialize(const char* buffer, const unsigned length) {
        // Clear scanner state variables here
        ...
        if (length > 0) {
            // rehydrate Scanner state variables here
            ...
        }
    }
```
I wrote my `deserialize` function this way copying the function from the [python scanner](https://github.com/tree-sitter/tree-sitter-python/blob/d6210ceab11e8d812d4ab59c07c81458ec6e5184/src/scanner.cc#L122) while forgetting to add the section clearing all the state variables. Hope this helps someone else avoid the same fate. Resetting your Scanner variables right at the top of your `deserialize` function is probably best practice no matter what, because you might implement your rehydration in a weird way that keeps some old state around.

**Reply by dfgordon** [reply]
Yes I made the very same mistake, so thanks indeed for posting.  Once you abstract this into a `Scanner` object it is easy to slip into the erroneous expectation that each time you scan you have a new object.  In the part of the docs about `deserialize` it could be worth adding a sentence such as "Remember your object is only created once, so you cannot rely on, e.g., the object's constructor to reset your variables."  True this is already implied, but apparently some of us missed the implication.
